Add support for HEOS groups

klada opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This adds support for grouping in the HEOS component by implementing the new `join`/`unjoin` services in the `media_player` component (#41193).

The PR addresses the following community feature request:
https://community.home-assistant.io/t/heos-group-support/126501/63


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository] (see: https://github.com/home-assistant/home-assistant.io/pull/16442).

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

klada committed:
Add support for grouping HEOS media players

klada commented:
@andrewsayre Feedback is welcome. I'll also take care of the HASS documentation, but I'd like to know if this approach is okay in general (especially the signature of the service calls).

andrewsayre reviewed (changes requested):
Thanks for your work on this -- I know `Group` support is a highly desired feature.  The issue here is that this functionality was jammed into the player entity when it should be represented as a separate Group entity entirely because the media controls are different.  I.e. you can still control the volume of the lead player independently than the group's volume.  You will need to refactor the group into a new entity type and support dynamic creation and removal as those are not static things.

klada commented:
@andrewsayre Group volume adjustment could actually be handled by a service call I think, which could be addressed by a separate PR. The overall approach is based on the way other integrations such as Bluesound and Sonos handle grouping of their speakers. That's why I just pushed out some changes to match their API (didn't think of that before). I also kind of dislike the service calls...

At first I also thought we need separate group entities for each HEOS group. But this would change the implementation and API significantly from other multi-room speaker integrations. And HEOS groups are very dynamic, they are created and destroyed on demand. So you could never actually address a group entity in Lovelace UI. The only static and addressable things in the HEOS system are the actual speakers.

The main benefit of the current approach is that existing media player cards such as mini-media-player (https://github.com/kalkih/mini-media-player) can work with HEOS groups without any code changes.

So what do you think? Do you really want to handle groups differently than other integrations?

andrewsayre commented:
I think groups should be a first-class item, despite how other integrations may have implemented it (that doesn't make it right).  Service calls are fine for joining and unjoining, but for basic operations, such as volume and playback control, that should be through media player HEOS group entity class.  Players (speakers) are transient too as they can be removed/joined to the network at any time, such as if the user hits the power button on their speaker.  The group entity itself can expose all the playback controls as the lead player, which should make it compatible with mini-media-player.  I know there's been discussion in the past on how to properly handle groups (and that we weren't happy with how it was done in Sonos).  Perhaps we open an arch issue to address this topic to get alignment on the path forward.

klada commented:
@andrewsayre I agree with you in that we should not make bad decisions only due to the fact that others made the same bad decisions. Compatibility with existing solutions (such as the popular mini-media-player) should still be taken into account when thinking of a solution.

Group entities sound like a clean approach. But personally I cannot really imagine how I (from a user perspective) would actually interact with them. Sometimes I want to play different things on all of my speakers. Sometimes I want to group the bedroom and bathroom speaker to one group and the living room and the kitchen speaker to another group. Sometimes I want all of my speakers to be in the same group.

From a frontend perspective, how could these dynamic groups be set up in Lovelace? Since they are dynamic there is no way of referencing them anywhere. Sure you could pre-create the same amount of dummy groups as there are speakers in the system (`media_player.heos_dummy_group_1` …  `media_player.heos_dummy_group_n`), but that is just... well... silly.

The only thing I can think of is artificially limiting the amount of HEOS groups to `1` and have a single HEOS group entity. But that would be a step back compared to the current approach. Or we could make the maximum number of HEOS groups configurable through `configuration.yaml`. Kind of clumsy, but hey...

Opening an arch issue sounds like a good idea. I think this needs more people to be involved in the discussion. Who is able to do that and where/how?

klada commented:
Ping @andrewsayre 
What is needed to convert this into an arch issue?

benganellison commented:
Please review the code. Waiting for this for some time. Don't stale it

klada commented:
> Please review the code. Waiting for this for some time. Don't stale it

I know many people are waiting for this. It's still on my todo list. I should be able to get some free time to finish the PR within the next two weeks. I have rebased it for now to remove the stale label.

klada commented:
@andrewsayre I have updated the PR to match the proposal of the architectural issue (https://github.com/home-assistant/architecture/issues/364).

I think we can change Sonos and Bluesound later, let's make use of this slightly modified grouping API in the HEOS component first.

klada reviewed (commented):
@andrewsayre This is now ready for review. I have changed my original implementation according to the suggestions in home-assistant/architecture#364

The constants which are supposed to be shared across the different multi-room audio compenents are now defined in `media_player/const.py`.

andrewsayre commented:
Great, I'll take a look within the next couple days.  Thanks!

andrewsayre commented on homeassistant/components/heos/media_player.py:
```diff
@@ -107,6 +120,7 @@ async def _player_update(self, player_id, event):
     async def _heos_updated(self):
         """Handle sources changed."""
         await self.async_update_ha_state(True)
+        self._group = await self.async_get_group()
```
This is redundant as `async_update_ha_state(True)` will call `async_update()` and that already performs this logic.

andrewsayre commented on homeassistant/components/heos/media_player.py:
```diff
@@ -122,12 +136,66 @@ async def async_added_to_hass(self):
                 SIGNAL_HEOS_UPDATED, self._heos_updated
             )
         )
+        self._group = await self.async_get_group()
```
This should not be called here.

andrewsayre commented on homeassistant/components/heos/media_player.py:
```diff
@@ -122,12 +136,66 @@ async def async_added_to_hass(self):
+    async def async_get_group(self):
+        """Build a dictionary with HEOS group membership information."""
+        group_membership = []
+        if not self.entity_id:
```
This condition should never occur when only called from within `async_update()`

klada replied:
OK, I removed the check.

andrewsayre commented on homeassistant/components/heos/media_player.py:
```diff
@@ -122,12 +136,66 @@ async def async_added_to_hass(self):
+        group_membership = []
+        if not self.entity_id:
+            return group_membership
+        if not self.player_id:
```
This condition is not possible.

klada replied:
OK, I removed the check.

andrewsayre commented on homeassistant/components/heos/media_player.py:
```diff
@@ -122,12 +136,66 @@ async def async_added_to_hass(self):
+            return group_membership
+
+        controller = self.hass.data[HEOS_DOMAIN][DATA_CONTROLLER_MANAGER].controller
+        if controller.connection_state != heos_const.STATE_CONNECTED:
```
This entire check and block of code should be removed.  The entity should not be checking the connection status directly.

klada replied:
OK, I removed the check.

andrewsayre commented on homeassistant/components/heos/media_player.py:
```diff
@@ -122,12 +136,66 @@ async def async_added_to_hass(self):
+            ):
+                leader_entity_id = None
+                for entity in self.hass.data[DOMAIN].entities:
+                    if not isinstance(entity, HeosMediaPlayer):
```
This can be added to the iterator:
```python
entity for entity in self.hass.data[DOMAIN].entities if isinstance(entity, HeosMediaPlayer):
```

klada replied:
@andrewsayre Sorry, I don't get this one. Could you please provide an example on how to add this to the current code? Either as an inline suggestion directly in this conversation or as a maintainer edit? Thanks.

andrewsayre commented on homeassistant/components/heos/media_player.py:
```diff
@@ -122,12 +136,66 @@ async def async_added_to_hass(self):
+                        leader_entity_id = entity.entity_id
+                    elif entity.player_id in member_player_ids:
+                        group_membership.append(entity.entity_id)
+                if not leader_entity_id:
```
Why would this ever occur?

klada replied:
In my tests it has never occured, so I have removed the check.

andrewsayre commented on homeassistant/components/heos/media_player.py:
```diff
@@ -122,12 +136,66 @@ async def async_added_to_hass(self):
+                    )
+                    return []
+                # Make sure the group leader is always the first element
+                group_membership = [leader_entity_id] + group_membership
```
```suggestion
                group_membership.insert(0, leader_entity_id)
```

andrewsayre commented on homeassistant/components/heos/media_player.py:
```diff
@@ -122,12 +136,66 @@ async def async_added_to_hass(self):
+                    return []
+                # Make sure the group leader is always the first element
+                group_membership = [leader_entity_id] + group_membership
+                _LOGGER.debug(
```
This isn't necessary to log.  It wasn't "discovered" either.

klada replied:
OK, I have removed the debug logging.

andrewsayre commented on homeassistant/components/heos/media_player.py:
```diff
@@ -122,12 +136,66 @@ async def async_added_to_hass(self):
+                )
+                return group_membership
+
+        _LOGGER.debug("No HEOS groups found for %s", self.entity_id)
```
Unnecessary logging

andrewsayre commented on homeassistant/components/heos/media_player.py:
```diff
@@ -270,6 +339,14 @@ def device_state_attributes(self) -> dict:
+            ATTR_GROUP_MEMBERS: self._group,
+            # All players which can be used for grouping. Right now these are all other
+            # HEOS devices.
+            ATTR_GROUP_CANDIDATES: [
```
This should be sourced from the available players in the API, not what's in HA as a`HeosMediaPlayer` instance.

andrewsayre commented on homeassistant/components/heos/services.py:
```diff
@@ -71,3 +104,138 @@ async def _sign_out_handler(controller, service):
+
+async def _join_handler(controller, hass, service):
+    """Join a HEOS player to a group."""
+    if controller.connection_state != const.STATE_CONNECTED:
```
When moved to an entity service, this check is no longer necessary.

andrewsayre commented on homeassistant/components/heos/services.py:
```diff
@@ -71,3 +104,138 @@ async def _sign_out_handler(controller, service):
         _LOGGER.error("Unable to sign out: %s", err)
+
+
+async def _join_handler(controller, hass, service):
```
This is overly complex.  `create_group` and `update_group` are the same thing in the API (they both translate to `set_group` behind the scenes.  It's fine to call `update_group` directly and pass in the IDs directly after resolving.  We shouldn't be checking existing groups and merging.  If the user intends to update an existing group, they should supply the entire group definition, otherwise you break the ability to reduce members.

klada replied:
Reducing members could also be done through the `unjoin()` service call, but you are right about the complexity. The new `HeosMediaPlayer.async_join()` method is quite simple now with the merging logic removed.

andrewsayre commented on homeassistant/components/heos/services.py:
```diff
@@ -71,3 +104,138 @@ async def _sign_out_handler(controller, service):
+
+async def _unjoin_handler(controller, hass, service):
+    """Unjoin HEOS players."""
+    if controller.connection_state != const.STATE_CONNECTED:
```
When moved to an entity service, this check is no longer necessary.

andrewsayre commented on homeassistant/components/heos/services.py:
```diff
@@ -71,3 +104,138 @@ async def _sign_out_handler(controller, service):
+            )
+
+
+async def _unjoin_handler(controller, hass, service):
```
This is also unnecessarily complex.  `remove_group` also calls `set_group` under the covers.  By passing a single `player_id`, it will ungroup--whether it was a member or the leader.  Simply call this method after resolving the id.

klada replied:
Good catch, I have updated the code in `HeosMediaPlayer.async_unjoin()`.

andrewsayre commented on homeassistant/components/heos/services.yaml:
```diff
@@ -10,3 +10,20 @@ sign_in:
+join:
+  description: Group players together as a HEOS group.
+  fields:
+    leader:
```
The entity doesn't represent the leader separately than the members, so why does the service call?  These should work the same way, otherwise it will confuse people of when to include the leader as the first element or when not to.

klada replied:
Yeah, I think this is obsolete now with the new `media_player.join` service.

andrewsayre reviewed (changes requested):
Thanks for your hard work on this.  There are some issues to fix and other items to clean up before this can move forward.  There is a drift from what @balloob described in this [comment from the architectural issue](https://github.com/home-assistant/architecture/issues/364#issuecomment-613603708).  
1. The `join` and `unjoin` services should be defined on the `media_player` entity component and then implemented in the entity class - not as integration services.
2. We need to be consistent between the service parameters and the attributes exposed by the entity.  Currently, the state attributes only list members (inclusive of the leader), but the leader is broken out in the service parameters.
3. The behavior of `join` and `unjoin` needs to be clarified.  Right now it's doing something different than how the underlying API works and results in a lot of complexity and unexpected behavior.  Join should accept a list of players (regardless of current membership) and pass that to the underlying service as the desired group.  HEOS handles whether to update or create a new one based on the leader.
4. Tests need to be added for the changes to the `media_player` integration as well as `heos`.

klada commented:
Thanks a lot for the review. :+1:  I'll have a look at it when I have some spare time.

> 1. The `join` and `unjoin` services should be defined on the `media_player` entity component and then implemented in the entity class - not as integration services.

This makes sense.

> 2. We need to be consistent between the service parameters and the attributes exposed by the entity.  Currently, the state attributes only list members (inclusive of the leader), but the leader is broken out in the service parameters.

Yeah, the leader is always the first element in the `members` list. That's the current state of Sonos and Bluesound and it's also the proposal from https://github.com/home-assistant/architecture/issues/364#issuecomment-625405413. But when thinking about it we could actually hide the leader concept entirely (not all components might have it anyway) and make it an implementation detail of the platform (first element is the leader, hide it in service calls).
 
> 3. The behavior of `join` and `unjoin` needs to be clarified.  Right now it's doing something different than how the underlying API works and results in a lot of complexity and unexpected behavior.  Join should accept a list of players (regardless of current membership) and pass that to the underlying service as the desired group.  HEOS handles whether to update or create a new one based on the leader.

That's quite different from the current approach, but it also makes sense. I like the idea of simplifying the `join` service call and hiding that leader concept. But we need to be careful to not push complexity to the automations. I need to think about possible use cases here first. What needs to be avoided is that joining additional players will create a new group as we don't want to interrupt the playback in an existing group.

> 4. Tests need to be added for the changes to the `media_player` integration as well as `heos`.

Okay, I'll see what I can do here.

Thanks again for the feedback.

klada commented on homeassistant/components/heos/services.py:
```diff
@@ -24,27 +32,52 @@
     )
+    hass.services.async_register(
+        HEOS_DOMAIN,
+        SERVICE_JOIN,
```
@andrewsayre : Your suggestion makes perfect sense. I guess you are talking about an `async_register_entity_service` call in `media_player`, just like this:

```python
component.async_register_entity_service(
    SERVICE_JOIN,
    {vol.Required(ATTR_ENTITY_ID): list},
    "async_join",
    [SUPPORT_GROUPING],
)
```

This would group all the players specified in `entity_id` together, hiding the leader concept. While this is a very nice and clean API (I really like it) it cannot be implemented like this because:

- This definition would call `async_join` on all players in `entity_id`
- The method `async_join` would not receive any arguments

I hope you see where my problem is here. We could implement it this way:

```python
component.async_register_entity_service(
    SERVICE_JOIN,
    {vol.Required(ATTR_GROUP_MEMBERS): list},
    "async_join",
    [SUPPORT_GROUPING],
)
```

Which means we'll end up with two arguments in the service call again: `entity_id` and `group_members`. For HEOS this means that `entity_id` will always be the leader and `group_membets` will be ... the group members. Would this be ok?

klada replied:
Ok, I decided split `entity_id` from `group_members` in the service call. This way it can be implemented as an entity service.

chrismdann commented:
Is this moving forward.

klada commented:
@andrewsayre Sorry, this one took me some time as my priorities shifted a little away from Home Assistant in the past weeks and months due to private reasons. I'd still like to give the code back to the community which I've been using in production ever since opening this PR.

I have updated the PR and resolved all but one of your change requests. Please see my comment above as I am a little stuck there.

I know I am still missing the test-part which seems to be required prior to merging this. Before I am putting any time into this I'd like to know if the rest of my PR is worth it or if there are some other blockers? If the rest of the code looks fine for you I'll try to take half a day off to finish it.

chrismdann commented:
Thank you for the work.  Is there a way for me to assist with testing.  I have heos but no testing experience.

klada commented:
@andrewsayre I really need some additional feedback here, please see my comment from above and that one unresolved inline comment. If you are too busy ATM it would be nice if you could assign this PR to another developer. Or we can meet on Discord or somewhere else for a realtime discussion, if necessary.

andrewsayre commented:
@klada I'll take a look early next week (I'm traveling this weekend)! 😃

klada commented:
I have split out the changes to the HASS core into a separate PR (#41193) which will hopefully make review easier. I'll rebase this HEOS PR as soon as the core PR has been accepted.

I know the discussion in this PR has become kind of lengthy (sorry about this) so to sum things up, this is the current state:

- I still don't know what to do about [this unresolved review comment](https://github.com/home-assistant/core/pull/32568#discussion_r459526433)
- I'll be looking into adding tests as soon as the core PR has been accepted and Andrew says that the rest of this HEOS PR looks mergable

keenedge commented:
@klada   I am following your work on this with great interest.  I'm looking for ways to contribute to home assistant.  I've been a programmer for many years but just getting up to speed on python development.  If you need testers I am ready help.

klada commented:
> @klada I am following your work on this with great interest. I'm looking for ways to contribute to home assistant. I've been a programmer for many years but just getting up to speed on python development. If you need testers I am ready help.

@keenedge  Thanks for offering your help. I won't be able to make any big contributions in the next ~8-10 weeks due to private reasons, but these are the things that need to be done:

- Someone needs to get a core developer to review #41193. Right now it looks like nobody is looking into that.
- If you want to help with testing, writing unit tests for the implementation in this PR would be a good start.

klada commented:
It looks like the core join/unjoin feature is about to be merged (#41193).  It was reviewed and approved by Paulus.  

@andrewsayre I have updated the implementation of this for HEOS and this time I have also added tests for it. Could you please look over the changes again (especially that one unresolved review comment)? Thanks.

**Please only merge if #41193 has been merged.**

klada commented:
@andrewsayre Could you please have a look at this? #41193 has been merged so this is ready now as well.

andrewsayre commented:
@klada Thanks for all your work on this. I'll review it this weekend! Appreciate your patience!

andrewsayre commented on homeassistant/components/heos/media_player.py:
```diff
@@ -235,10 +302,26 @@ async def async_update(self):
+    @log_command_error("unjoin_player")
+    async def async_unjoin_player(self):
+        """Remove this player from any group."""
+        if self.group_members:
```
I would remove this check and simply call the API command. This will ensure the player is ungrouped should there have been an error previously populating the group members.

klada replied:
You are right, removing that check makes it more robust.

andrewsayre commented on homeassistant/components/heos/media_player.py:
```diff
@@ -235,10 +302,26 @@ async def async_update(self):
+        if self.group_members:
+            controller = self.hass.data[HEOS_DOMAIN][DATA_CONTROLLER_MANAGER].controller
+            try:
+                await controller.create_group(self.player_id, [])
```
```suggestion
                await controller.remove_group(self.player_id)
```

klada replied:
`remove_group` is different, it requires a group id (which we do not have and we actually don't need it). I think you have pointed out in your first review that passing a single `player_id` to `set_group`/`create_group` properly removes any group for that player.

andrewsayre commented on homeassistant/components/heos/media_player.py:
```diff
@@ -128,6 +136,65 @@ async def async_clear_playlist(self):
+            await controller.create_group(self.player_id, player_ids)
+        except HeosError as err:
+            _LOGGER.error(
+                "HEOS create_group with arguments (%s %s) failed. Error: %s",
```
Don't leak the underlying API and parameters into user-facing error messages. Instead, log something that is understandable to a user and reference the entity IDs: 
```suggestion
                "Failed to group %s with %s: %s",
```

andrewsayre commented on homeassistant/components/heos/media_player.py:
```diff
@@ -128,6 +136,65 @@ async def async_clear_playlist(self):
+        except HeosError as err:
+            _LOGGER.error(
+                "HEOS create_group with arguments (%s %s) failed. Error: %s",
+                self.player_id,
```
```suggestion
                self.entity_id,
```

andrewsayre commented on homeassistant/components/heos/media_player.py:
```diff
@@ -128,6 +136,65 @@ async def async_clear_playlist(self):
+            _LOGGER.error(
+                "HEOS create_group with arguments (%s %s) failed. Error: %s",
+                self.player_id,
+                player_ids,
```
```suggestion
                group_members,
```

andrewsayre commented on homeassistant/components/heos/media_player.py:
```diff
@@ -235,10 +302,26 @@ async def async_update(self):
+                await controller.create_group(self.player_id, [])
+            except HeosError as err:
+                _LOGGER.error(
+                    "HEOS create_group with arguments (%s %s) failed. Error: %s",
```
Follow the same change as requested above:

```suggestion
                    "Failed to ungroup %s: %s",
```

andrewsayre commented on homeassistant/components/heos/media_player.py:
```diff
@@ -235,10 +302,26 @@ async def async_update(self):
+            except HeosError as err:
+                _LOGGER.error(
+                    "HEOS create_group with arguments (%s %s) failed. Error: %s",
+                    self.player_id,
```
```suggestion
                    self.entity_id,
```

andrewsayre commented on homeassistant/components/heos/media_player.py:
```diff
@@ -235,10 +302,26 @@ async def async_update(self):
+                _LOGGER.error(
+                    "HEOS create_group with arguments (%s %s) failed. Error: %s",
+                    self.player_id,
+                    [],
```
Remove this

andrewsayre commented on homeassistant/components/heos/media_player.py:
```diff
@@ -235,10 +302,26 @@ async def async_update(self):
         controls = self._player.now_playing_media.supported_controls
         current_support = [CONTROL_TO_SUPPORT[control] for control in controls]
         self._supported_features = reduce(ior, current_support, BASE_SUPPORTED_FEATURES)
+        self._group = await self.async_get_group()
```
The logic to retrieve group definitions is controller-scoped (not player-scoped) and should follow the same [pattern used to maintain sources](https://github.com/home-assistant/core/blob/dev/homeassistant/components/heos/__init__.py#L220-L344) by introducing a `group_manager` stored in `self.hass.data[HEOS_DOMAIN][DATA_GROUP_MANAGER]`. This manager should subscribe to `EVENT_GROUPS_CHANGED` and trigger the state updates in the dependent entities. It should also encapsulate all of the logic to resolve HEOS `player_ids` to HASS `entity_ids`.

klada replied:
Good point. As you have suggested, I have moved the code to the `GroupManager` class which handles the fetching of group information for all players based on events.

andrewsayre reviewed (changes requested):
Thanks for all of your work on this. It's getting closer!

The biggest change needed is related to how groups are loaded. This is a controller-level resource currently being refreshed at the player-level. This means _each player_ calls this function to refresh _all_ of the groups every time _each player_ is updated. Groups only need to be retrieved at startup or when they change (which the current approach doesn't handle). To fix this, follow the same pattern as the [SourceManager](https://github.com/home-assistant/core/blob/dev/homeassistant/components/heos/__init__.py#L220-L344). This is also a controller-level resource that loads at startup, refreshes when sources change, and triggers entity updates in the downstream players. This new `GroupManager` will maintain the list of groups, subscribe to `EVENT_GROUPS_CHANGED` and encapsulate the interactions with the API, including the translation of `player_ids` to/from `entity_ids`. `HeosMediaPlayer` will then use the `GroupManager` for all group-related operations (not the API directly).

Let me know if you have any questions,
Andy

klada commented:
Thanks for the review, this sounds like a good approach. I definitely need some time for these changes as my daily screen time is very limited due to health issues. If somebody wants to assist or take over I'll happily accept commits. But I'll see what I can do.

klada committed:
Update homeassistant/components/heos/media_player.py

Co-authored-by: Andrew Sayre (he/his/him) <6730289+andrewsayre@users.noreply.github.com>

klada committed:
Update homeassistant/components/heos/media_player.py

Co-authored-by: Andrew Sayre (he/his/him) <6730289+andrewsayre@users.noreply.github.com>

klada committed:
Update homeassistant/components/heos/media_player.py

Co-authored-by: Andrew Sayre (he/his/him) <6730289+andrewsayre@users.noreply.github.com>

klada committed:
Update homeassistant/components/heos/media_player.py

Co-authored-by: Andrew Sayre (he/his/him) <6730289+andrewsayre@users.noreply.github.com>

klada committed:
Update homeassistant/components/heos/media_player.py

Co-authored-by: Andrew Sayre (he/his/him) <6730289+andrewsayre@users.noreply.github.com>

klada committed:
Update homeassistant/components/heos/media_player.py

Co-authored-by: Andrew Sayre (he/his/him) <6730289+andrewsayre@users.noreply.github.com>

klada committed:
Handle groups at controller level, refine tests.

Co-authored-by: Andrew Sayre (he/his/him) <6730289+andrewsayre@users.noreply.github.com>

klada committed:
Merge branch 'dev' into heos_groups

klada committed:
Fix linting issues

klada commented on homeassistant/components/heos/media_player.py:
```diff
@@ -235,10 +302,26 @@ async def async_update(self):
+            controller = self.hass.data[HEOS_DOMAIN][DATA_CONTROLLER_MANAGER].controller
+            try:
+                await controller.create_group(self.player_id, [])
+            except HeosError as err:
```
Okay, I have covered the errors through the tests now.

klada commented:
@andrewsayre I have applied your suggestions now. Could you please have a look at this again? Thanks.

heikkih commented:
Hope to see this reviewed and approved soon (and that this comment will remove the stale label)

ajuch commented:
I'm interested in this feature, isn't there somebody else that can review the code? It would be a waste of effort if this code wasn't merged...

quizzical86 commented:
@klada has been very committed to implementing all the feedback he received, I commend his efforts and sincerely hope someone can review his work

klada commented:
@andrewsayre  I know time is a very limited resource for all of us. Is there still a chance you could have a look at this anytime soon? If not, could you please reassign this PR, because I think nobody will look at this as long as you are assigned to it? Thanks.

klada committed:
Update homeassistant/components/heos/media_player.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

klada committed:
Update homeassistant/components/heos/media_player.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

klada committed:
Rename variables and improve resolving of entity_ids

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

klada commented on homeassistant/components/heos/__init__.py:
```diff
@@ -219,6 +227,117 @@ def update_ids(self, mapped_ids: dict[int, int]):
+
+        return {
+            entity.entity_id: entity.player_id
+            for entity in self._hass.data[MEDIA_PLAYER_DOMAIN].entities
```
Yeah, this makes sense. Thanks for the suggestion :)

klada commented on homeassistant/components/heos/__init__.py:
```diff
@@ -219,6 +227,117 @@ def update_ids(self, mapped_ids: dict[int, int]):
+            if isinstance(entity, HeosMediaPlayer)
+        }
+
+    def _get_heos_ids_to_entities(self) -> dict:
```
True. I have also renamed all of the other variables where this applies.

klada committed:
Don't patch internal methods

Use the pytest fixtures which have already been defined for this.

klada committed:
Fix linting issues

klada commented on tests/components/heos/test_media_player.py:
```diff
@@ -805,3 +811,99 @@ async def test_play_media_invalid_type(hass, config_entry, config, controller, c
+    """Test group_members attribute."""
+    await setup_platform(hass, config_entry, config)
+    with patch.object(controller, "get_groups", return_value=group) as mock_get_groups:
+        hass.data[DOMAIN][DATA_GROUP_MANAGER].force_update_groups()
```
You are right. I am not very familiar with pytest and it's fixture concept but apparently there really is no need to patch any internal methods. I have updated my tests so they are using the same concepts as the rest of the component (pytest fixtures).

klada commented:
@MartinHjelmare Thank you very much for taking the time to review this. I have applied your suggestions so this is ready for another review round :)

klada committed:
Remove unused property

MartinHjelmare commented on homeassistant/components/heos/__init__.py:
```diff
@@ -219,6 +227,115 @@ def update_ids(self, mapped_ids: dict[int, int]):
+            heos_const.SIGNAL_HEOS_EVENT, self.async_update_groups
+        )
+
+    def force_update_groups(self):
```
Why do we need this method?

klada replied:
The players need to call this when they have registered their `entity_id`. But since all other use-cases for calling this method have vanished now I guess it makes more sense to use a signal for this. So I have removed this method and the players are now sending `SIGNAL_HEOS_PLAYER_ADDED` which is handled by the group manager.

MartinHjelmare commented on homeassistant/components/heos/media_player.py:
```diff
@@ -125,12 +134,22 @@ async def async_added_to_hass(self):
+        self.hass.data[HEOS_DOMAIN][DATA_ENTITY_ID_MAP][
+            self._player.player_id
+        ] = self.entity_id
+        self._group_manager.force_update_groups()
```
We should call the async api from async context.

klada replied:
True. Since I have replaces the method with a signal I am now using `hass.helpers.dispatcher.async_dispatcher_send` right there instead. So we can keep the async context.

MartinHjelmare commented on tests/components/heos/test_media_player.py:
```diff
@@ -805,3 +810,93 @@ async def test_play_media_invalid_type(hass, config_entry, config, controller, c
+):
+    """Test group_members attribute."""
+    await setup_platform(hass, config_entry, config)
+    hass.data[DOMAIN][DATA_GROUP_MANAGER].force_update_groups()
```
Please don't access integration details in the test. Set up the integration and let it add the entity and update the group.

klada replied:
It turns out that after my last refactoring `force_update_groups ` doesn't have to be called at all during tests, as the player instances handle this. :smile:

MartinHjelmare commented on tests/components/heos/test_media_player.py:
```diff
@@ -805,3 +810,93 @@ async def test_play_media_invalid_type(hass, config_entry, config, controller, c
+    assert "Unable to get HEOS group info" not in caplog.text
+
+    controller.get_groups.side_effect = HeosError("error")
+    hass.data[DOMAIN][DATA_GROUP_MANAGER].force_update_groups()
```
Let the controller signal the manager to update instead.

klada replied:
It turns out that after my last refactoring `force_update_groups ` doesn't have to be called at all during tests, as the player instances handle this. :smile:

klada committed:
Ignore groups with unknown leader

This makes sure that the group_members attribute won't contain a `None`
value as a leader entity_id.

klada committed:
Don't call force_update_groups() from tests

klada commented on homeassistant/components/heos/__init__.py:
```diff
@@ -236,71 +236,63 @@ def __init__(self, hass, controller):
+        player_id_to_entity_id_map = self._hass.data[DOMAIN][DATA_ENTITY_ID_MAP]
         for group in groups.values():
-            leader_entity_id = id_to_entity.get(group.leader.player_id)
+            leader_entity_id = player_id_to_entity_id_map.get(group.leader.player_id)
```
You are right. I have added a check for a `None` leader.

MartinHjelmare replied:
Do we know when/why this could happen?

klada replied:
I actually only see this happening when the integration is still loading (when not all MediaPlayer entities have been loaded).

MartinHjelmare replied:
Do you think the current solution is good enough or do we need to wait calling this, maybe with asyncio synchronization, until all platforms are done loading?

klada replied:
The benefit of the current solution is that if new players are discovered at runtime their group information is also populated right when they appear (although it's unlikely that new players are part of any groups). The drawback is that we are calling `async_get_group_membership` once for every player at startup (which may cause these `None` entities for the first initialized `HeosMediaPlayer` for a tiny amount of time).

Is it possible to use something like the `component_loaded` event in the integration? In this case I could wait until this event is fired and resolve groups afterwards. As a fallback I guess a global "home_assistant_loaded" event would also suffice. Unknown/`None` entity_ids would then be quite unlikely in either case.

MartinHjelmare replied:
If we forward the platform setup tasks inside another task we can gather and await the platform forward tasks and after that set an `asyncio.Event`. The event can then be used for synchronization where needed.

klada replied:
I now used a much simpler way of synchronization which prevents unnecessary calls to `async_get_group_membership` during startup :)

klada committed:
Don't pass `None` player ids to HEOS API

klada commented on homeassistant/components/heos/__init__.py:
```diff
@@ -219,6 +227,115 @@ def update_ids(self, mapped_ids: dict[int, int]):
+        entity_id_to_player_id_map = self._get_entity_id_to_player_id_map()
+        leader_id = entity_id_to_player_id_map.get(leader_entity_id)
+        member_ids = [
+            entity_id_to_player_id_map.get(member) for member in member_entity_ids
```
This would actually not really matter as we'd get an error from the controller in this case (which is properly handled). But I have now created a check for `None` ids so we don't pass invalid data to the underlying library on purpose.

klada committed:
Use signal for group manager communication

MartinHjelmare commented on homeassistant/components/heos/__init__.py:
```diff
@@ -331,9 +333,13 @@ def connect_update(self):
+        async def _async_handle_player_added():
+            await self.async_update_groups(SIGNAL_HEOS_PLAYER_ADDED)
+
+        self._hass.helpers.dispatcher.async_dispatcher_connect(
```
Please import the `async_dispatcher_connect` instead of using it on `hass.helpers`. The latter is legacy.

klada replied:
Thanks for pointing that out. I am importing `async_dispatcher_connect`/`async_dispatcher_send` now.

MartinHjelmare commented on homeassistant/components/heos/__init__.py:
```diff
@@ -219,6 +228,125 @@ def update_ids(self, mapped_ids: dict[int, int]):
+                self._group_membership = groups
+                _LOGGER.debug("Groups updated due to change event")
+                # Let players know to update
+                self._hass.helpers.dispatcher.async_dispatcher_send(SIGNAL_HEOS_UPDATED)
```
Please import `async_dispatcher_send` instead. We prefer to import to make typing work.

klada replied:
Thanks for pointing that out. I am importing `async_dispatcher_connect`/`async_dispatcher_send` now.

MartinHjelmare commented on homeassistant/components/heos/__init__.py:
```diff
@@ -219,6 +228,125 @@ def update_ids(self, mapped_ids: dict[int, int]):
+            if member in entity_id_to_player_id_map
+        ]
+
+        if leader_id:
```
We should probably raise an exception here if there's no leader_id, with the message as exception argument. The message will be shown to the user in the frontend.

klada replied:
Alright, I am now raising `HomeAssistantError`. Invalid input (invalid entity_ids) will usually not make it's way down to this method as it's filtered out by the service call `media_player.join`/`media_player.unjoin`.

MartinHjelmare commented on homeassistant/components/heos/__init__.py:
```diff
@@ -219,6 +228,125 @@ def update_ids(self, mapped_ids: dict[int, int]):
+        """Remove `player_entity_id` from any group."""
+        player_id = self._get_entity_id_to_player_id_map().get(player_entity_id)
+
+        if player_id:
```
If there's no player id we should raise an exception with a message as argument. If the error is due to invalid input we raise `ValueError` other `HomeAssistantError`.

klada replied:
Alright, I am now raising `HomeAssistantError`. Invalid input (invalid entity_ids) will usually not make it's way down to this method as it's filtered out by the service call `media_player.join`/`media_player.unjoin`.

klada committed:
Use imports for async_dispatcher_send/async_dispatcher_connect

klada committed:
Raise exception when leader/player could not be resolved

MartinHjelmare commented on homeassistant/components/heos/__init__.py:
```diff
@@ -337,8 +347,8 @@ def connect_update(self):
 
-        self._hass.helpers.dispatcher.async_dispatcher_connect(
-            SIGNAL_HEOS_PLAYER_ADDED, _async_handle_player_added
+        async_dispatcher_connect(
```
Do we need to disconnect this signal when reloading the config entry?

klada replied:
Is `async_unload_entry` called on reaload? In this case I could just disconnect the signal right there.

klada replied:
Alright, the signal is now disconnected in `disconnect_update()` through `entry.async_on_unload()`

klada committed:
Disconnect signal handlers, avoid calling async_update_groups too early

MartinHjelmare commented on homeassistant/components/heos/__init__.py:
```diff
@@ -219,6 +233,147 @@ def update_ids(self, mapped_ids: dict[int, int]):
+            if (
+                len(self._hass.data[DOMAIN][MEDIA_PLAYER_DOMAIN])
+                <= len(self._hass.data[DOMAIN][DATA_ENTITY_ID_MAP])
+                and not self._initialized
```
Is it only needed to update once per config entry setup?

klada replied:
Yes, this is just required during initial config entry setup. In all other cases the players signal updated groups through the HEOS-native event `heos_const.EVENT_CONNECTED`.

And in case the config entry is reloaded a new `GroupManager` instance will be created which will be initialized the very same way.

klada committed:
Update homeassistant/components/heos/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

MartinHjelmare reviewed (approved):
Looks good!

klada commented:
@MartinHjelmare  Thank you very much for the time you have put into the review. I have learned a lot about Home Assistant internals through the hints you gave me.

heikkih commented:
🎉 
Thanks a lot for the great work on getting this in, @klada and @MartinHjelmare. Can't wait to start playing with automations on this now.
😄
